pkg/private/zip/build_zip.py

```diff
@@ -68,53 +69,145 @@ def parse_date(ts):
+
+    # See: https://trac.edgewall.org/attachment/ticket/8919/ZipDownload.patch
+    # external_attr is 4 bytes in size. The high order two bytes represent UNIX
+    # permission and file type bits, while the low order two contain MS-DOS FAT
+    # file attributes.
+    if mode:
+      f_mode = int(mode, 8)
+    else:
+      f_mode = self.default_mode
+    entry_info.external_attr = f_mode << 16
+    return entry_info
+
+  def add_manifest_entry(self, options, entry):
+    """Add an entry to the zip file.
+
+    Args:
+      options: parsed options
+      zip_file: ZipFile to write to
+      entry: manifest entry
+    """
+
+    entry_type, dest, src, mode, user, group = entry
+
+    # Use the pkg_tar mode/owner remaping as a fallback
+    non_abs_path = dest.strip('/')
+    dst_path = _combine_paths(options.directory, non_abs_path)
+    if entry_type == manifest.ENTRY_IS_DIR and not dst_path.endswith('/'):
+      dst_path += '/'
+    entry_info = self.make_zipinfo(path=dst_path, mode=mode)
+
+    if entry_type == manifest.ENTRY_IS_FILE:
+      entry_info.compress_type = zipfile.ZIP_DEFLATED
+      with open(src, 'rb') as src:
+        self.zip_file.writestr(entry_info, src.read())
+    elif entry_type == manifest.ENTRY_IS_DIR:
+      entry_info.compress_type = zipfile.ZIP_STORED
+      # Set directory bits
+      entry_info.external_attr |= (UNIX_DIR_BIT << 16) | MSDOS_DIR_BIT
+      self.zip_file.writestr(entry_info, '')
+    elif entry_type == manifest.ENTRY_IS_LINK:
+      entry_info.compress_type = zipfile.ZIP_STORED
+      # Set directory bits
+      entry_info.external_attr |= (UNIX_SYMLINK_BIT << 16)
+      self.zip_file.writestr(entry_info, src)
+    elif entry_type == manifest.ENTRY_IS_TREE:
+      self.add_tree(src, dst_path, mode)
+    elif entry.entry_type == manifest.ENTRY_IS_EMPTY_FILE:
+      entry_info.compress_type = zipfile.ZIP_DEFLATED
+      self.zip_file.writestr(entry_info, '')
+    else:
+      raise Exception('Unknown type for manifest entry:', entry)
+
+  def add_tree(self, tree_top: str, destpath: str, mode: int):
+    """Add a tree artifact to the zip file.
+
+    Args:
+      tree_top: the top of the tree to add
+      destpath: the path under which to place the files
+      mode: if not None, file mode to apply to all files
+    """
+
+    # We expect /-style paths.
+    tree_top = os.path.normpath(tree_top).replace(os.path.sep, '/')
+
+    # Again, we expect /-style paths.
+    dest = destpath.strip('/')  # redundant, dests should never have / here
+    dest = os.path.normpath(dest).replace(os.path.sep, '/')
+    # paths should not have a leading ./
+    dest = '' if dest == '.' else dest + '/'
+
+    to_write = {}
+    for root, dirs, files in os.walk(tree_top):
+      # While `tree_top` uses '/' as a path separator, results returned by
+      # `os.walk` and `os.path.join` on Windows may not.
+      root = os.path.normpath(root).replace(os.path.sep, '/')
+
+      dirs = sorted(dirs)
```

Comment:
Don't we need to create dir entries?

Reply:
That is a good question.

We don't have to. unzip will create them as needed.
And, since zip can not specify owners, and the directories should all  755, I don't see the point.

In the future, we *might* want the ability to let the mode specified on the tree artifact apply to intermediate directories, but I would rather do that when someone comes up with a good use case, and treat that as a separate feature requreest.

Reply:
Line removed.